File: easy/53_maximum_subarray.rb
# 53. Maximum Subarray
# Easy
#
# Given an integer array nums, find the contiguous subarray (containing at
# least one number) which has the largest sum and return its sum.

def max_sub_array(nums)
  cur = nums[0]
  max = nums[0]

  nums[1..].each do |n|
    if n > cur + n
      cur = n
    else
      cur += n
    end
    max = cur if cur > max
  end

  max
end

puts max_sub_array([-2, 1, -3, 4, -1, 2, 1, -5, 4]) # 6
puts max_sub_array([1])                             # 1
puts max_sub_array([5, 4, -1, 7, 8])                # 23

File: go/53_maximum_subarray.go
// 53. Maximum Subarray
// Medium
//
// Given an integer array nums, find the contiguous subarray (containing at
// least one number) which has the largest sum and return its sum.
//
// A subarray is a contiguous part of an array.
package main

func maxSubArray(nums []int) int {
	cur := nums[0]
	max := nums[0]

	for i := 1; i < len(nums); i++ {
		n := nums[i]
		if n > cur+n {
			cur = n
		} else {
			cur += n
		}
		if cur > max {
			max = cur
		}
	}

	return max
}
